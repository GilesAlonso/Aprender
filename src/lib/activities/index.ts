export * from "./types";
export * from "./catalog";
export * from "./quiz";
export * from "./puzzle";
export * from "./game";
