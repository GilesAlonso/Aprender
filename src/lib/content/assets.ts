import { readFileSync, existsSync } from "node:fs";
import path from "node:path";

import { z } from "zod";

export const EF01_ASSETS_DIR = path.join(process.cwd(), "public", "assets", "ef01");
export const EF01_MANIFEST_PATH = path.join(process.cwd(), "data", "assets", "ef01-manifest.json");

const ef01AssetEntrySchema = z.object({
  slug: z.string().min(1),
  subject: z.enum(["lingua-portuguesa", "ciencias"]),
  theme: z.string().min(1),
  recommendedUsage: z.string().min(1),
  altText: z.string().min(1),
  colorProfile: z.array(z.string().min(1)).nonempty(),
  license: z.object({
    name: z.string().min(1),
    url: z.string().url(),
    document: z.string().min(1),
  }),
  files: z.object({
    svg: z.string().min(1),
    png: z.string().min(1),
  }),
  hash: z.string().min(6),
});

const ef01ManifestSchema = z.object({
  version: z.string().min(1),
  generatedBy: z.string().min(1),
  assets: z.array(ef01AssetEntrySchema).nonempty(),
});

export type Ef01AssetManifestEntry = z.infer<typeof ef01AssetEntrySchema>;
export type Ef01AssetManifest = z.infer<typeof ef01ManifestSchema>;

export const loadEf01AssetManifest = (): Ef01AssetManifest => {
  const content = readFileSync(EF01_MANIFEST_PATH, "utf-8");
  const parsed = ef01ManifestSchema.parse(JSON.parse(content));

  for (const entry of parsed.assets) {
    const svgPath = path.join(
      process.cwd(),
      "public",
      entry.files.svg.split("?")[0].replace(/^\//, "")
    );
    const pngPath = path.join(
      process.cwd(),
      "public",
      entry.files.png.split("?")[0].replace(/^\//, "")
    );

    if (!existsSync(svgPath)) {
      throw new Error(
        `Arquivo SVG ausente para o asset EF01 '${entry.slug}' em '${entry.files.svg}'.`
      );
    }

    if (!existsSync(pngPath)) {
      throw new Error(
        `Arquivo PNG ausente para o asset EF01 '${entry.slug}' em '${entry.files.png}'.`
      );
    }
  }

  return parsed;
};

export const buildEf01AssetIndex = (
  manifest: Ef01AssetManifest
): Map<string, Ef01AssetManifestEntry> => {
  const index = new Map<string, Ef01AssetManifestEntry>();

  for (const entry of manifest.assets) {
    index.set(entry.slug, entry);
  }

  return index;
};
