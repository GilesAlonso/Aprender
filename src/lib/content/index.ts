export * from "./schema";
export * from "./workspace";
export * from "./lint";
export * from "./build";
export * from "./assets";
