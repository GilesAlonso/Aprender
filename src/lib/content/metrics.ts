import { difficultyTierSchema, type DifficultyTier } from "./schema";
import type { ContentModuleDocument } from "./workspace";

export type ContentTotals = {
  modules: number;
  activities: number;
};

export type StageSubjectMetrics = ContentTotals;

export type StageMetrics = ContentTotals & {
  subjects: Record<string, StageSubjectMetrics>;
};

export type BNCCMetrics = {
  codes: Record<string, number>;
  habilidades: Record<string, number>;
};

export type DifficultyDistribution = Record<DifficultyTier, number>;

export type ContentMetrics = {
  totals: ContentTotals;
  stages: Record<string, StageMetrics>;
  bncc: BNCCMetrics;
  difficulty: DifficultyDistribution;
};

type SubjectAccumulator = {
  modules: number;
  activities: number;
};

type StageAccumulator = {
  modules: number;
  activities: number;
  subjects: Map<string, SubjectAccumulator>;
};

const difficultyTiers = difficultyTierSchema.options;

const createDifficultyDistribution = (): DifficultyDistribution => {
  const distribution = {} as DifficultyDistribution;
  for (const tier of difficultyTiers) {
    distribution[tier] = 0;
  }
  return distribution;
};

const sortEntries = <T>(input: Iterable<[string, T]>): Array<[string, T]> =>
  Array.from(input).sort((a, b) => a[0].localeCompare(b[0]));

const toRecord = (entries: Iterable<[string, number]>): Record<string, number> => {
  const result: Record<string, number> = {};
  for (const [key, value] of sortEntries(entries)) {
    result[key] = value;
  }
  return result;
};

export const computeContentMetrics = (documents: ContentModuleDocument[]): ContentMetrics => {
  const totals: ContentTotals = {
    modules: documents.length,
    activities: 0,
  };

  const difficulty = createDifficultyDistribution();
  const bnccCodeCounts = new Map<string, number>();
  const habilidadeCounts = new Map<string, number>();
  const stageMap = new Map<string, StageAccumulator>();

  const getStageAccumulator = (stageSlug: string): StageAccumulator => {
    const normalized = stageSlug.toLowerCase();
    let stage = stageMap.get(normalized);
    if (!stage) {
      stage = {
        modules: 0,
        activities: 0,
        subjects: new Map<string, SubjectAccumulator>(),
      };
      stageMap.set(normalized, stage);
    }
    return stage;
  };

  const getSubjectAccumulator = (
    stage: StageAccumulator,
    subjectSlug: string
  ): SubjectAccumulator => {
    const normalized = subjectSlug.toLowerCase();
    let subject = stage.subjects.get(normalized);
    if (!subject) {
      subject = { modules: 0, activities: 0 };
      stage.subjects.set(normalized, subject);
    }
    return subject;
  };

  for (const document of documents) {
    const stageAccumulator = getStageAccumulator(document.module.stage);
    const subjectAccumulator = getSubjectAccumulator(stageAccumulator, document.module.subjectSlug);

    stageAccumulator.modules += 1;
    stageAccumulator.activities += document.activities.length;
    subjectAccumulator.modules += 1;
    subjectAccumulator.activities += document.activities.length;

    totals.activities += document.activities.length;

    for (const activity of document.activities) {
      difficulty[activity.difficulty] += 1;

      const bnccCode = activity.bncc.code;
      bnccCodeCounts.set(bnccCode, (bnccCodeCounts.get(bnccCode) ?? 0) + 1);

      for (const habilidade of activity.bncc.habilidades) {
        habilidadeCounts.set(habilidade, (habilidadeCounts.get(habilidade) ?? 0) + 1);
      }
    }
  }

  const stages: Record<string, StageMetrics> = {};

  for (const [stageSlug, accumulator] of sortEntries(stageMap)) {
    const subjects: Record<string, StageSubjectMetrics> = {};
    for (const [subjectSlug, subjectAccumulator] of sortEntries(accumulator.subjects)) {
      subjects[subjectSlug] = {
        modules: subjectAccumulator.modules,
        activities: subjectAccumulator.activities,
      };
    }

    stages[stageSlug] = {
      modules: accumulator.modules,
      activities: accumulator.activities,
      subjects,
    };
  }

  return {
    totals,
    stages,
    bncc: {
      codes: toRecord(bnccCodeCounts),
      habilidades: toRecord(habilidadeCounts),
    },
    difficulty,
  };
};
