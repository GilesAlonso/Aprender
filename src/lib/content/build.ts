import { createHash } from "node:crypto";
import { writeFileSync } from "node:fs";
import path from "node:path";

import type { ContentActivity, ContentModule } from "./schema";
import { buildEf01AssetIndex, loadEf01AssetManifest, type Ef01AssetManifestEntry } from "./assets";
import { loadWorkspace, CONTENT_DIR } from "./workspace";

export type BuildOptions = {
  rootDir?: string;
  writeToDisk?: boolean;
  stage?: string;
  subjectSlug?: string;
  modules?: string[];
};

export type ActivityIndexEntry = {
  moduleSlug: string;
  activitySlug: string;
  activityType: string;
};

export type ModulesDataset = {
  updatedAt: string;
  hash: string;
  modules: ContentModuleOutput[];
};

export type InteractiveDataset = {
  updatedAt: string;
  hash: string;
  activities: InteractiveActivityOutput[];
};

export type ContentIndexDataset = {
  generatedAt: string;
  contentHash: string;
  modulesBySubject: Record<string, string[]>;
  modulesByStage: Record<string, string[]>;
  activitiesByDifficulty: Record<string, ActivityIndexEntry[]>;
  interactiveMap: Record<string, ActivityIndexEntry>;
  bncc: {
    modules: Record<string, string>;
    activities: Record<string, string>;
  };
};

export type BuildOutput = {
  modules: ModulesDataset;
  interactive: InteractiveDataset;
  index: ContentIndexDataset;
};

type Ef01AssetIndex = Map<string, Ef01AssetManifestEntry>;

const difficultyToLegacy: Record<string, string> = {
  INICIAR: "BEGINNER",
  PRATICAR: "INTERMEDIATE",
  DOMINAR: "ADVANCED",
};

const isPlainObject = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const normalizeHints = (value: unknown): Array<Record<string, unknown>> | undefined => {
  if (!Array.isArray(value)) {
    return undefined;
  }

  const normalized = value
    .map((hint) => {
      if (typeof hint === "string") {
        return { text: hint } as Record<string, unknown>;
      }

      if (isPlainObject(hint)) {
        return hint as Record<string, unknown>;
      }

      return null;
    })
    .filter((hint): hint is Record<string, unknown> => hint !== null);

  return normalized.length > 0 ? normalized : undefined;
};

const removeEmpty = (value: Record<string, unknown>): Record<string, unknown> | undefined => {
  const entries = Object.entries(value).filter(([, entryValue]) => {
    if (entryValue === undefined || entryValue === null) {
      return false;
    }

    if (Array.isArray(entryValue)) {
      return entryValue.length > 0;
    }

    if (isPlainObject(entryValue)) {
      return Object.keys(entryValue).length > 0;
    }

    return true;
  });

  return entries.length > 0 ? Object.fromEntries(entries) : undefined;
};

type AccessibilityValue = NonNullable<ContentActivity["accessibility"]>;

type AccessibilityContext = {
  moduleSlug: string;
  subjectSlug: string;
  activitySlug: string;
  scope: "activity" | "interactive";
  interactiveSlug?: string;
};

const normalizeSupportEntry = (entry: unknown): Record<string, unknown> | null => {
  if (typeof entry === "string") {
    const normalized = entry.trim();
    return normalized.length > 0 ? { description: normalized } : null;
  }

  if (isPlainObject(entry)) {
    const descriptionValue = entry["description"];
    const titleValue = entry["title"];
    const stepsValue = entry["steps"];

    const description =
      typeof descriptionValue === "string" && descriptionValue.trim().length > 0
        ? descriptionValue
        : undefined;

    if (!description) {
      return null;
    }

    const steps = Array.isArray(stepsValue)
      ? stepsValue
          .map((item) => (typeof item === "string" ? item.trim() : String(item)))
          .filter((item) => item.length > 0)
      : undefined;

    return (
      removeEmpty({
        title:
          typeof titleValue === "string" && titleValue.trim().length > 0 ? titleValue : undefined,
        description,
        steps,
      }) ?? null
    );
  }

  return null;
};

const normalizeSupportCollection = (value: unknown): Record<string, unknown>[] | undefined => {
  if (value === undefined || value === null) {
    return undefined;
  }

  const items = Array.isArray(value) ? value : [value];
  const normalized = items
    .map((entry) => normalizeSupportEntry(entry))
    .filter((entry): entry is Record<string, unknown> => entry !== null);

  return normalized.length > 0 ? normalized : undefined;
};

const describeAccessibilityContext = (context: AccessibilityContext): string => {
  if (context.scope === "interactive") {
    return `interativo '${context.interactiveSlug ?? "desconhecido"}' da atividade '${context.activitySlug}' no módulo '${context.moduleSlug}'`;
  }

  return `atividade '${context.activitySlug}' no módulo '${context.moduleSlug}'`;
};

const normalizeAccessibility = (
  accessibility: ContentActivity["accessibility"],
  assetIndex: Ef01AssetIndex,
  context: AccessibilityContext
): Record<string, unknown> | undefined => {
  if (!accessibility) {
    return undefined;
  }

  if (!isPlainObject(accessibility)) {
    return undefined;
  }

  const data = accessibility as Record<string, unknown>;

  const hints = normalizeHints(data["hints"]);

  const feedbackRaw = data["feedback"];
  const feedback = isPlainObject(feedbackRaw)
    ? (feedbackRaw as Record<string, unknown>)
    : undefined;

  const prerequisitesRaw = data["prerequisites"];
  const prerequisites =
    Array.isArray(prerequisitesRaw) && prerequisitesRaw.length > 0
      ? (prerequisitesRaw as unknown[])
      : undefined;

  const assetsRaw = data["assets"];

  const assets =
    Array.isArray(assetsRaw) && assetsRaw.length > 0
      ? (assetsRaw as AccessibilityValue["assets"])
          .map((asset) => {
            const base: Record<string, unknown> = {
              type: asset.type,
              slug: asset.slug,
              uri: asset.uri,
              title: asset.title,
              description: asset.description,
              altText: asset.altText,
            };

            if (asset.slug) {
              const manifestEntry = assetIndex.get(asset.slug);
              if (!manifestEntry) {
                const location = describeAccessibilityContext(context);
                throw new Error(
                  `Asset '${asset.slug}' referenciado em ${location} não existe no manifest EF01.`
                );
              }

              const resolvedUri = asset.uri ?? manifestEntry.files.svg;
              const normalized = removeEmpty({
                ...base,
                subject: manifestEntry.subject,
                theme: manifestEntry.theme,
                uri: resolvedUri,
                png: manifestEntry.files.png,
                altText: asset.altText ?? manifestEntry.altText,
                recommendedUsage: manifestEntry.recommendedUsage,
                colorProfile: manifestEntry.colorProfile,
                hash: manifestEntry.hash,
                license: manifestEntry.license,
              });

              if (!normalized || !normalized["uri"]) {
                const location = describeAccessibilityContext(context);
                throw new Error(`Asset '${asset.slug}' em ${location} não pôde ser normalizado.`);
              }

              return normalized;
            }

            if (!asset.uri) {
              const location = describeAccessibilityContext(context);
              throw new Error(
                `Asset sem slug nem URI encontrado em ${location}. Verifique o conteúdo.`
              );
            }

            if (asset.type === "IMAGEM" && !asset.altText) {
              const location = describeAccessibilityContext(context);
              throw new Error(
                `Asset de imagem '${asset.uri}' em ${location} precisa informar 'altText' ou referenciar um slug válido.`
              );
            }

            return removeEmpty(base) ?? undefined;
          })
          .filter(
            (value): value is Record<string, unknown> =>
              value !== undefined && Object.keys(value).length > 0
          )
      : undefined;

  const alternatives = normalizeSupportCollection(data["alternatives"]);
  const audioCueCollection = normalizeSupportCollection(data["audioCue"]);
  const audioCue = audioCueCollection ? audioCueCollection[0] : undefined;
  const motorSupport = normalizeSupportCollection(data["motorSupport"]);

  return (
    removeEmpty({
      hints,
      feedback,
      assets,
      prerequisites,
      alternatives,
      audioCue,
      motorSupport,
    }) ?? undefined
  );
};

const mapDifficulty = (difficulty: string): string => difficultyToLegacy[difficulty] ?? difficulty;

const computeHash = (value: unknown): string => {
  const hash = createHash("sha256");
  hash.update(JSON.stringify(value));
  return hash.digest("hex");
};

type ModuleActivityOutput = {
  slug: string;
  title: string;
  prompt: string;
  activityType: string;
  difficulty: string;
  bnccCode: string;
  learningObjectives: string[];
  description?: string;
  metadata?: Record<string, unknown>;
  summary?: string;
};

type ContentModuleOutput = {
  slug: string;
  stage: string;
  subjectSlug: string;
  subject: string;
  title: string;
  subtitle?: string;
  description: string;
  theme?: string;
  discipline?: string;
  ageGroupSlug: string;
  learningPathSlug?: string;
  primaryBnccCode: string;
  secondaryBnccCodes?: string[];
  learningOutcomes: string[];
  tags?: string[];
  estimatedDurationMinutes?: number;
  summary?: string;
  notes?: string[];
  activities: ModuleActivityOutput[];
};

type InteractiveActivityOutput = {
  slug: string;
  title: string;
  type: string;
  bnccCode: string;
  bnccDescription: string;
  contentModuleSlug: string;
  activitySlug: string;
  difficulty: string;
  estimatedTimeMinutes: number;
  instructions: string[];
  objectives: string[];
  accessibility?: Record<string, unknown>;
  quiz?: Record<string, unknown>;
  puzzle?: Record<string, unknown>;
  game?: Record<string, unknown>;
};

const buildActivityMetadata = (
  activity: ContentActivity,
  module: ContentModule,
  assetIndex: Ef01AssetIndex
): Record<string, unknown> | undefined => {
  const base: Record<string, unknown> = activity.metadata ? { ...activity.metadata } : {};

  base.bncc = {
    code: activity.bncc.code,
    habilidades: activity.bncc.habilidades,
  };

  if (activity.habilidadesDescription) {
    base.habilidadesDescription = activity.habilidadesDescription;
  }

  base.difficultyTier = activity.difficulty;

  const accessibility = normalizeAccessibility(activity.accessibility, assetIndex, {
    moduleSlug: module.slug,
    subjectSlug: module.subjectSlug,
    activitySlug: activity.slug,
    scope: "activity",
  });

  if (accessibility) {
    base.accessibility = accessibility;
  }

  if (activity.prerequisites && activity.prerequisites.length > 0) {
    base.prerequisites = activity.prerequisites;
  }

  if (activity.notes && activity.notes.length > 0) {
    base.notes = activity.notes;
  }

  if (activity.interactive) {
    base.interactiveSlug = activity.interactive.slug;
    base.interactiveType = activity.interactive.type;
  }

  return removeEmpty(base);
};

const buildInteractiveEntry = (
  module: ContentModule,
  activity: ContentActivity,
  difficulty: string,
  assetIndex: Ef01AssetIndex
): InteractiveActivityOutput | null => {
  if (!activity.interactive) {
    return null;
  }

  const { interactive } = activity;
  const base: InteractiveActivityOutput = {
    slug: interactive.slug,
    title: interactive.title,
    type: interactive.type,
    bnccCode: activity.bncc.code,
    bnccDescription: interactive.bnccDescription,
    contentModuleSlug: module.slug,
    activitySlug: activity.slug,
    difficulty,
    estimatedTimeMinutes: interactive.estimatedTimeMinutes,
    instructions: [...interactive.instructions],
    objectives: [...interactive.objectives],
  };

  const accessibility = normalizeAccessibility(interactive.accessibility, assetIndex, {
    moduleSlug: module.slug,
    subjectSlug: module.subjectSlug,
    activitySlug: activity.slug,
    scope: "interactive",
    interactiveSlug: interactive.slug,
  });

  if (accessibility) {
    base.accessibility = accessibility;
  }

  if (interactive.type === "QUIZ") {
    base.quiz = {
      ...(interactive.scoring ? { scoring: interactive.scoring } : {}),
      ...(interactive.adaptive ? { adaptive: interactive.adaptive } : {}),
      questions: interactive.questions,
    };
  } else if (interactive.type === "PUZZLE") {
    base.puzzle = interactive.puzzle;
  } else if (interactive.type === "GAME") {
    base.game = interactive.game;
  }

  return base;
};

const buildModuleEntry = (
  module: ContentModule,
  activities: ModuleActivityOutput[]
): ContentModuleOutput => ({
  slug: module.slug,
  stage: module.stage,
  subjectSlug: module.subjectSlug,
  subject: module.subject,
  title: module.title,
  subtitle: module.subtitle,
  description: module.description,
  theme: module.theme,
  discipline: module.discipline,
  ageGroupSlug: module.ageGroupSlug,
  learningPathSlug: module.learningPathSlug ?? undefined,
  primaryBnccCode: module.primaryBnccCode,
  secondaryBnccCodes: module.secondaryBnccCodes,
  learningOutcomes: module.learningOutcomes,
  tags: module.tags,
  estimatedDurationMinutes: module.estimatedDurationMinutes,
  summary: module.summary,
  notes: module.notes,
  activities,
});

const buildModuleActivityEntry = (
  activity: ContentActivity,
  module: ContentModule,
  assetIndex: Ef01AssetIndex
): ModuleActivityOutput => {
  const difficulty = mapDifficulty(activity.difficulty);
  const metadata = buildActivityMetadata(activity, module, assetIndex);

  return {
    slug: activity.slug,
    title: activity.title,
    prompt: activity.prompt,
    activityType: activity.type,
    difficulty,
    bnccCode: activity.bncc.code,
    learningObjectives: activity.learningObjectives,
    description: activity.description,
    metadata,
    summary: activity.summary,
  };
};

export const buildWorkspace = (options: BuildOptions = {}): BuildOutput => {
  const documents = loadWorkspace({
    rootDir: options.rootDir,
    stage: options.stage,
    subjectSlug: options.subjectSlug,
    modules: options.modules,
  });
  const ef01Manifest = loadEf01AssetManifest();
  const ef01AssetIndex = buildEf01AssetIndex(ef01Manifest);
  const buildTimestamp = new Date().toISOString();

  const modulesBySubject: Record<string, Set<string>> = {};
  const modulesByStage: Record<string, Set<string>> = {};
  const activitiesByDifficulty: Record<string, ActivityIndexEntry[]> = {
    INICIAR: [],
    PRATICAR: [],
    DOMINAR: [],
  };
  const interactiveIndex: Record<string, ActivityIndexEntry> = {};
  const bnccModulesIndex: Record<string, string> = {};
  const bnccActivitiesIndex: Record<string, string> = {};

  const modulesOutput: ContentModuleOutput[] = [];
  const interactiveOutput: InteractiveActivityOutput[] = [];

  for (const document of documents) {
    bnccModulesIndex[document.module.slug] = document.module.primaryBnccCode;

    if (!modulesBySubject[document.module.subjectSlug]) {
      modulesBySubject[document.module.subjectSlug] = new Set();
    }
    modulesBySubject[document.module.subjectSlug].add(document.module.slug);

    if (!modulesByStage[document.module.stage]) {
      modulesByStage[document.module.stage] = new Set();
    }
    modulesByStage[document.module.stage].add(document.module.slug);

    const activityEntries = document.activities.map((activity) => {
      bnccActivitiesIndex[activity.slug] = activity.bncc.code;

      const difficulty = mapDifficulty(activity.difficulty);
      activitiesByDifficulty[activity.difficulty] =
        activitiesByDifficulty[activity.difficulty] ?? [];
      activitiesByDifficulty[activity.difficulty].push({
        moduleSlug: document.module.slug,
        activitySlug: activity.slug,
        activityType: activity.type,
      });

      const interactiveEntry = buildInteractiveEntry(
        document.module,
        activity,
        difficulty,
        ef01AssetIndex
      );
      if (interactiveEntry) {
        interactiveOutput.push(interactiveEntry);
        interactiveIndex[interactiveEntry.slug] = {
          moduleSlug: document.module.slug,
          activitySlug: activity.slug,
          activityType: activity.type,
        };
      }

      return buildModuleActivityEntry(activity, document.module, ef01AssetIndex);
    });

    modulesOutput.push(buildModuleEntry(document.module, activityEntries));
  }

  modulesOutput.sort((a, b) => a.slug.localeCompare(b.slug));
  interactiveOutput.sort((a, b) => a.slug.localeCompare(b.slug));

  const modulesDatasetWithoutHash = {
    updatedAt: buildTimestamp,
    modules: modulesOutput,
  };
  const modulesHash = computeHash(modulesDatasetWithoutHash);
  const modulesDataset: ModulesDataset = {
    ...modulesDatasetWithoutHash,
    hash: modulesHash,
  };

  const interactiveDatasetWithoutHash = {
    updatedAt: buildTimestamp,
    activities: interactiveOutput,
  };
  const interactiveHash = computeHash(interactiveDatasetWithoutHash);
  const interactiveDataset: InteractiveDataset = {
    ...interactiveDatasetWithoutHash,
    hash: interactiveHash,
  };

  const modulesBySubjectSorted = Object.fromEntries(
    Object.entries(modulesBySubject)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([subject, set]) => [subject, Array.from(set).sort((a, b) => a.localeCompare(b))])
  );

  const modulesByStageSorted = Object.fromEntries(
    Object.entries(modulesByStage)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([stage, set]) => [stage, Array.from(set).sort((a, b) => a.localeCompare(b))])
  );

  const activitiesByDifficultySorted = Object.fromEntries(
    Object.entries(activitiesByDifficulty).map(([difficulty, entries]) => [
      difficulty,
      entries
        .slice()
        .sort((a, b) =>
          a.moduleSlug === b.moduleSlug
            ? a.activitySlug.localeCompare(b.activitySlug)
            : a.moduleSlug.localeCompare(b.moduleSlug)
        ),
    ])
  );

  const interactiveMapSorted = Object.fromEntries(
    Object.entries(interactiveIndex).sort(([a], [b]) => a.localeCompare(b))
  );

  const bnccModulesSorted = Object.fromEntries(
    Object.entries(bnccModulesIndex).sort(([a], [b]) => a.localeCompare(b))
  );

  const bnccActivitiesSorted = Object.fromEntries(
    Object.entries(bnccActivitiesIndex).sort(([a], [b]) => a.localeCompare(b))
  );

  const indexDatasetWithoutHash: ContentIndexDataset = {
    generatedAt: buildTimestamp,
    contentHash: "",
    modulesBySubject: modulesBySubjectSorted,
    modulesByStage: modulesByStageSorted,
    activitiesByDifficulty: activitiesByDifficultySorted,
    interactiveMap: interactiveMapSorted,
    bncc: {
      modules: bnccModulesSorted,
      activities: bnccActivitiesSorted,
    },
  };

  const contentHash = computeHash({
    modules: modulesDatasetWithoutHash,
    interactive: interactiveDatasetWithoutHash,
    index: indexDatasetWithoutHash,
  });

  const indexDataset: ContentIndexDataset = {
    ...indexDatasetWithoutHash,
    contentHash,
  };

  if (options.writeToDisk ?? true) {
    const modulesPath = path.join(CONTENT_DIR, "modules.json");
    const interactivePath = path.join(CONTENT_DIR, "interactive-activities.json");
    const indexPath = path.join(CONTENT_DIR, "index.json");

    writeFileSync(modulesPath, `${JSON.stringify(modulesDataset, null, 2)}\n`);
    writeFileSync(interactivePath, `${JSON.stringify(interactiveDataset, null, 2)}\n`);
    writeFileSync(indexPath, `${JSON.stringify(indexDataset, null, 2)}\n`);
  }

  return {
    modules: modulesDataset,
    interactive: interactiveDataset,
    index: indexDataset,
  };
};
