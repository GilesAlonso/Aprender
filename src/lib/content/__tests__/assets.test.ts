import { existsSync, readFileSync } from "node:fs";
import path from "node:path";

import { describe, expect, it } from "vitest";

import { loadEf01AssetManifest } from "@/lib/content";

describe("EF01 asset manifest", () => {
  const manifest = loadEf01AssetManifest();

  it("garante que cada entrada possui arquivos e descrição alternativa", () => {
    expect(manifest.assets.length).toBeGreaterThan(0);

    for (const asset of manifest.assets) {
      expect(asset.altText.trim().length).toBeGreaterThan(0);
      expect(asset.files.svg).toContain(".svg");
      expect(asset.files.png).toContain(".png");

      const svgPath = asset.files.svg.split("?")[0].replace(/^\//, "");
      const pngPath = asset.files.png.split("?")[0].replace(/^\//, "");

      expect(existsSync(path.join(process.cwd(), "public", svgPath))).toBe(true);
      expect(existsSync(path.join(process.cwd(), "public", pngPath))).toBe(true);
    }
  });

  it("mantém SVGs determinísticos para amostras representativas", () => {
    const sampleSlugs = ["lp-silaba-ba", "cn-animal-tucano"] as const;

    const samples = sampleSlugs.reduce<Record<string, string>>((accumulator, slug) => {
      const entry = manifest.assets.find((asset) => asset.slug === slug);
      expect(entry, `Asset '${slug}' precisa existir no manifest`).toBeDefined();
      if (!entry) {
        return accumulator;
      }

      const svgFile = entry.files.svg.split("?")[0].replace(/^\//, "");
      const svgContent = readFileSync(path.join(process.cwd(), "public", svgFile), "utf-8").trim();
      accumulator[slug] = svgContent;
      return accumulator;
    }, {});

    expect(samples).toMatchSnapshot();
  });
});
