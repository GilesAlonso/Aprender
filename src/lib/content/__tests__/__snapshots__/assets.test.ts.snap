// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EF01 asset manifest > mantém SVGs determinísticos para amostras representativas 1`] = `
{
  "cn-animal-tucano": "<svg xmlns="http://www.w3.org/2000/svg" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" width="256" height="256" viewBox="0 0 256 256"><rect width="256" height="256" fill="#b7ceff" rx="48" ry="48"></rect><rect width="220" height="20" rx="10" ry="10" fill="#e88500" x="18" y="190" transform="matrix(0.9986295347545738,-0.05233595624294383,0.05233595624294383,0.9986295347545738,-10.291771697174227,6.973095448182022)"></rect><g><ellipse rx="75" ry="60" cx="128" cy="144" fill="#19387f"></ellipse><circle r="35" cx="174" cy="84" fill="#19387f"></circle><ellipse rx="34" ry="16" cx="218" cy="68" fill="#ff9f1a" transform="matrix(0.9902680687415704,-0.13917310096006544,0.13917310096006544,0.9902680687415704,-7.3422098509468015,31.001507334867483)"></ellipse><circle r="8" cx="152" cy="88" fill="#fdfdfd"></circle><circle r="4" cx="156" cy="88" fill="#1e293b"></circle><path d="M138 168 Q154 196 170 168" stroke-linecap="round" stroke-width="8" stroke="#0f172a" fill="none"></path></g><ellipse rx="41" ry="23" cx="50" cy="164" fill="#0fa958" transform="matrix(0.9271838545667874,-0.374606593415912,0.374606593415912,0.9271838545667874,-57.79467404854894,30.672177521842485)"></ellipse><ellipse rx="29" ry="15" cx="30" cy="196" fill="#2abf80" transform="matrix(0.9510565162951535,-0.3090169943749474,0.3090169943749474,0.9510565162951535,-59.09902638634429,18.863432637398347)"></ellipse></svg>",
  "lp-silaba-ba": "<svg xmlns="http://www.w3.org/2000/svg" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" width="256" height="256" viewBox="0 0 256 256"><rect width="256" height="256" fill="#ffe4c3" rx="48" ry="48"></rect><rect width="208" height="160" fill="#ff8d1f" rx="36" ry="36" x="24" y="60"></rect><text font-family="'Baloo 2', 'Nunito', 'Inter', sans-serif" font-size="120" text-anchor="middle" font-weight="700" fill="#fdfdfd" x="128" y="192.552734375" data-svgjs="{&quot;leading&quot;:&quot;1&quot;}"><tspan dy="0" x="128">BA</tspan></text><rect width="72" height="20" fill="#ffb6eb" rx="10" ry="10" x="92" y="26"></rect></svg>",
}
`;
